<a href="https://colab.research.google.com/github/kwansawanth/BSC_DPDM23/blob/main/Copy_of_Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
import pandas as pd
import os

In [9]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/bsc2 2566/bsc_dpdm23_data'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [10]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [11]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

##ข้อมูลมาจากกี่ประเทศ

In [12]:
data.groupby(['Country']).sum()

<ipython-input-12-ca3bc1201a7e>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  data.groupby(['Country']).sum()


Quantity    UnitPrice    CustomerID
Country                                                  
Australia                83653     4054.750  1.569300e+07
Austria                   4827     1701.520  5.021102e+06
Bahrain                    260       86.570  2.100270e+05
Belgium                  23152     7540.130  2.571829e+07
Brazil                     356      142.600  4.086080e+05
Canada                    2763      910.580  2.615483e+06
Channel Islands           9479     3738.550  1.128522e+07
Cyprus                    6317     3920.070  7.715880e+06
Czech Republic             592       88.150  3.834300e+05
Denmark                   8188     1266.950  4.876734e+06
EIRE                    142637    48447.190  1.103917e+08
European Community         497      294.050  9.215880e+05
Finland                  10666     3786.850  8.699324e+06
France                  110480    43031.990  1.076489e+08
Germany                 117448    37666.000  1.200751e+08
Greece                    1556      713.290  2.008584e+06
Hong Kong                 4769    12241.500  0.000000e+00
Iceland                   2458      481.210  2.247154e+06
Israel                    4353     1079.040  3.164467e+06
Italy                     7999     3879.390  1.015666e+07
Japan                    25218      814.860  4.567292e+06
Lebanon                    386      242.440  5.743800e+05
Lithuania                  652       99.440  5.366200e+05
Malta                      944      666.010  2.158496e+06
Netherlands             200128     6492.550  3.419054e+07
Norway                   19247     6529.060  1.350765e+07
Poland                    3653     1422.270  4.341972e+06
Portugal                 16180    13037.540  1.886480e+07
RSA                        352      248.100  7.218680e+05
Saudi Arabia                75       24.110  1.256500e+05
Singapore                 5234    25108.890  2.918376e+06
Spain                    26824    12633.450  3.268929e+07
Sweden                   35637     1806.830  6.790083e+06
Switzerland              30325     6813.690  2.377592e+07
USA                       1034      644.980  3.672086e+06
United Arab Emirates       982      229.890  1.018952e+06
United Kingdom         4263829  2245715.474  5.626433e+09
Unspecified               3300     1204.010  3.348046e+06

#แปลงdata ให้อยู่ในรูปtransaction

In [13]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [14]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [15]:
!pip install apyori

In [16]:
from apyori import apriori

In [17]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [18]:
len(frequentItemsets)

4

In [19]:
frequentItemsets[0][0]

frozenset({'22423'})

In [20]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [21]:
frequentItemsets[0][0].__len__()

1

In [22]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [23]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

In [24]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'544893',
 '557117',
 '579711',
 '554039',
 'C567358',
 '542527',
 '542719',
 '547830',
 '563584',
 '578149',
 '576329',
 '552149',
 '539609',
 '553013',
 '547493',
 '546849',
 '566603',
 '574561',
 'C545724',
 '544095',
 '564975',
 '570834',
 '561217',
 '548983',
 '561891',
 '544205',
 '554842',
 '575846',
 'C545438',
 '550205',
 '556205',
 '539752',
 '556237',
 '580612',
 '564758',
 '567671',
 '560357',
 '538371',
 '557500',
 '559173',
 '569650',
 '566600',
 '575138',
 '539451',
 '547936',
 '575947',
 '565724',
 '553541',
 '562374',
 '556083',
 '539739',
 '549927',
 '572092',
 '548003',
 '544305',
 '554798',
 '536946',
 '576063',
 '545063',
 '546669',
 '555150',
 '565779',
 '546158',
 '562157',
 '555594',
 '570701',
 '554110',
 '576246',
 '543533',
 '560225',
 '577258',
 '558923',
 '564172',
 '556636',
 '570187',
 '553870',
 '541423',
 '558315',
 '558904',
 '549192',
 '550613',
 '562113',
 '537176',
 '540992',
 '570677',
 '568655',
 '565214',
 '538566',
 '559679',
 '572751',
 '55356

#Quiz สินค้าอะไร

In [25]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B')]['Description'])

{'JUMBO BAG PINK POLKADOT', 'JUMBO BAG RED RETROSPOT'}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

#Hw1 หา Frequent set item set และ Association rules


*   ของช่วงเวลา Christmas & New year (1 dec - 30 dec)
*   ของประเทศยุโรป
*   ของช่วงเวลาตรุษจีนในทวีป Asia ( 15 Jan - 15 Feb)
*   ของช่วงเวลา Christmas & New year Europe (1 dec - 30 dec)
*   Valentine's day
*   โดยจัดกลุ่มตามประเทศ
groupby ประเทศ - เดือน








###ข้อ1 ช่วงเวลา Christmas and New year (1 Dec - 31 Dec)

In [26]:
import datetime

In [27]:
data['InvoiceDate']

0         12/01/2010 08:26
1         12/01/2010 08:26
2         12/01/2010 08:26
3         12/01/2010 08:26
4         12/01/2010 08:26
                ...       
541904    12/09/2011 12:50
541905    12/09/2011 12:50
541906    12/09/2011 12:50
541907    12/09/2011 12:50
541908    12/09/2011 12:50
Name: InvoiceDate, Length: 541909, dtype: object

In [28]:
#data['InvoiceDate']=pd.to_datetime(data.InvoiceDate, format='%m/%d/%Y')

In [29]:
data['InvoiceDate']=pd.to_datetime(data.InvoiceDate, format='%m/%d/%Y %H:%M' )

In [30]:
data['Date'] = data['InvoiceDate'].dt.date

In [31]:
data['Date']=pd.to_datetime(data.Date, format='%Y/%m/%d')

In [32]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country       Date  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom 2010-12-01  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom 2010-12-01  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
...                    ...        ...         ...             ...        ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France 2011-12-09  
541905 2011-12-09 12:50:00       2.10     12680.0          France 2011-12-09  
541906 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541907 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541908 2011-12-09 12:50:00       4.95     12680.0          France 2011-12-09  

[541909 rows x 9 columns]

In [33]:
#Select 1-31 Dec
December_1 = data[(data['Date'].dt.month == 12) & (data['Date'].dt.day >= 1) & (data['Date'].dt.day <= 31)]
#ปริ้นเพื่อตรวจสอบ
print(December_1)


       InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country       Dat

In [34]:
transactions1 = December_1.groupby('InvoiceNo')['Description'].apply(list).reset_index()
print(transactions1)

     InvoiceNo                                        Description
0       536365  [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1       536366  [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2       536367  [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3       536368  [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4       536369                         [BATH BUILDING BLOCK WORD]
...        ...                                                ...
3035   C581484                      [PAPER CRAFT , LITTLE BIRDIE]
3036   C581490  [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
3037   C581499                                           [Manual]
3038   C581568                       [VICTORIAN SEWING BOX LARGE]
3039   C581569  [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...

[3040 rows x 2 columns]


In [35]:
transactions1_apply = transactions1['Description'].apply(lambda x: [str(item) for item in x])
transactions1_apply

0       [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1       [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2       [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3       [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4                              [BATH BUILDING BLOCK WORD]
                              ...                        
3035                        [PAPER CRAFT , LITTLE BIRDIE]
3036    [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
3037                                             [Manual]
3038                         [VICTORIAN SEWING BOX LARGE]
3039    [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...
Name: Description, Length: 3040, dtype: object

In [36]:
frequentItemset = list(apriori(transactions1_apply, min_support=0.036))

for record in frequentItemset:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKELIKE RED 
Support: 0.0368 (3.68%)
Rule 1: (None -> ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKELIKE RED  : Confidence=0.0368, Lift=1.0000)
Rule 2: (ALARM CLOCK BAKELIKE GREEN -> ALARM CLOCK BAKELIKE RED  : Confidence=0.7089, Lift=14.4627)
Rule 3: (ALARM CLOCK BAKELIKE RED  -> ALARM CLOCK BAKELIKE GREEN : Confidence=0.7517, Lift=14.4627)

yes
Items: CHOCOLATE HOT WATER BOTTLE, HOT WATER BOTTLE I AM SO POORLY
Support: 0.0365 (3.65%)
Rule 1: (None -> CHOCOLATE HOT WATER BOTTLE, HOT WATER BOTTLE I AM SO POORLY : Confidence=0.0365, Lift=1.0000)
Rule 2: (CHOCOLATE HOT WATER BOTTLE -> HOT WATER BOTTLE I AM SO POORLY : Confidence=0.5163, Lift=9.7484)
Rule 3: (HOT WATER BOTTLE I AM SO POORLY -> CHOCOLATE HOT WATER BOTTLE : Confidence=0.6894, Lift=9.7484)

yes
Items: PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS
Support: 0.0477 (4.77%)
Rule 1: (None -> PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS : 

###ข้อ2 ของประเทศในทวีปยุโรป

In [37]:
europe = ['Austria' ,'Belgium' ,'Channel Islands' ,'Cyprus' ,'Czech Republic' ,'Denmark' , 'EIRE' ,'European Community' , 'Finland' ,'France' ,'Germany' , 'Greece' , 'Iceland' , 'Italy' , 'Lithuania' , 'Malta' ,'Netherlands' , 'Norway' ,'Poland' ,'Portugal' , 'Spain' ,'Sweden' ,'Switzerland' ,'United Kingdom']

In [38]:
EU_data = data[data['Country']==""]
for i in range(len(europe)):
  x = data[data['Country']==europe[i]]
  EU_data=pd.concat([x,EU_data])

In [39]:
print(EU_data)

       InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

               InvoiceDate  UnitPrice  CustomerID         Country       Dat

In [40]:
transactions2 = EU_data.groupby('InvoiceNo')['Description'].apply(list).reset_index()
print(transactions2)

      InvoiceNo                                        Description
0        536365  [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1        536366  [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2        536367  [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3        536368  [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4        536369                         [BATH BUILDING BLOCK WORD]
...         ...                                                ...
25726   C581484                      [PAPER CRAFT , LITTLE BIRDIE]
25727   C581490  [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
25728   C581499                                           [Manual]
25729   C581568                       [VICTORIAN SEWING BOX LARGE]
25730   C581569  [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...

[25731 rows x 2 columns]


In [41]:
transactions2_apply = transactions2['Description'].apply(lambda x: [str(item) for item in x])
transactions2_apply

0        [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1        [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2        [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3        [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4                               [BATH BUILDING BLOCK WORD]
                               ...                        
25726                        [PAPER CRAFT , LITTLE BIRDIE]
25727    [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
25728                                             [Manual]
25729                         [VICTORIAN SEWING BOX LARGE]
25730    [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...
Name: Description, Length: 25731, dtype: object

In [42]:
frequentItemset2 = list(apriori(transactions2_apply, min_support=0.025))

for record in frequentItemset2:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()

yes
Items: ROSES REGENCY TEACUP AND SAUCER , GREEN REGENCY TEACUP AND SAUCER
Support: 0.0301 (3.01%)
Rule 1: (None -> ROSES REGENCY TEACUP AND SAUCER , GREEN REGENCY TEACUP AND SAUCER : Confidence=0.0301, Lift=1.0000)
Rule 2: (GREEN REGENCY TEACUP AND SAUCER -> ROSES REGENCY TEACUP AND SAUCER  : Confidence=0.7402, Lift=17.2054)
Rule 3: (ROSES REGENCY TEACUP AND SAUCER  -> GREEN REGENCY TEACUP AND SAUCER : Confidence=0.7001, Lift=17.2054)

yes
Items: JUMBO BAG RED RETROSPOT, JUMBO BAG PINK POLKADOT
Support: 0.0323 (3.23%)
Rule 1: (None -> JUMBO BAG RED RETROSPOT, JUMBO BAG PINK POLKADOT : Confidence=0.0323, Lift=1.0000)
Rule 2: (JUMBO BAG PINK POLKADOT -> JUMBO BAG RED RETROSPOT : Confidence=0.6775, Lift=8.2039)
Rule 3: (JUMBO BAG RED RETROSPOT -> JUMBO BAG PINK POLKADOT : Confidence=0.3915, Lift=8.2039)

yes
Items: JUMBO BAG RED RETROSPOT, JUMBO SHOPPER VINTAGE RED PAISLEY
Support: 0.0265 (2.65%)
Rule 1: (None -> JUMBO BAG RED RETROSPOT, JUMBO SHOPPER VINTAGE RED PAISLEY : Confidence=0

###ข้อ 3 ของช่วงเวลาตรุษจันในทวีปเอเชีย 15Jan-15Feb

In [43]:
Asia = ['Bahrain' , 'Hong Kong' , 'Israel' , 'Japan' , 'Lebanon' , 'Saudi Arabia' , 'Singapore', 'United Arab Emirates']

In [44]:
Asia_data = data[data['Country']==""]
for i in range(len(Asia)):
  x = data[data['Country']==Asia[i]]
  Asia_data=pd.concat([x,Asia_data])

In [45]:
Asia_data

InvoiceNo StockCode                         Description  Quantity  \
89570     543911     21485    RETROSPOT HEART HOT WATER BOTTLE         6   
89571     543911     48138                  DOORMAT UNION FLAG         2   
89572     543911     85150       LADIES & GENTLEMEN METAL SIGN        24   
89573     543911     82552                 WASHROOM METAL SIGN        24   
89574     543911     82551              LAUNDRY 15C METAL SIGN        12   
...          ...       ...                                 ...       ...   
181152    552449    72802C   VANILLA SCENT CANDLE JEWELLED BOX         6   
197371    553900     37449  CERAMIC CAKE STAND + HANGING CAKES         2   
197372    553900     37446  MINI CAKE STAND WITH HANGING CAKES         8   
197373    553900     22890  NOVELTY BISCUITS CAKE STAND 3 TIER         4   
197374    553900     37450   CERAMIC CAKE BOWL + HANGING CAKES         6   

               InvoiceDate  UnitPrice  CustomerID               Country  \
89570  2011-02-14 12:46:00       4.95     17829.0  United Arab Emirates   
89571  2011-02-14 12:46:00       7.95     17829.0  United Arab Emirates   
89572  2011-02-14 12:46:00       2.55     17829.0  United Arab Emirates   
89573  2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
89574  2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
...                    ...        ...         ...                   ...   
181152 2011-05-09 13:49:00       4.25     12355.0               Bahrain   
197371 2011-05-19 17:47:00       9.95     12353.0               Bahrain   
197372 2011-05-19 17:47:00       1.45     12353.0               Bahrain   
197373 2011-05-19 17:47:00       9.95     12353.0               Bahrain   
197374 2011-05-19 17:47:00       2.95     12353.0               Bahrain   

             Date  
89570  2011-02-14  
89571  2011-02-14  
89572  2011-02-14  
89573  2011-02-14  
89574  2011-02-14  
...           ...  
181152 2011-05-09  
197371 2011-05-19  
197372 2011-05-19  
197373 2011-05-19  
197374 2011-05-19  

[1314 rows x 9 columns]

In [46]:
#Chinese Newyear
China1 = Asia_data[(Asia_data['Date'].dt.month == 1) & (Asia_data['Date'].dt.day >= 15) & (data['Date'].dt.day <= 31)  |
                   ((Asia_data['Date'].dt.month == 2) & (Asia_data['Date'].dt.day <= 15))]
#ปริ้นเพื่อตรวจสอบ
print(China1)

      InvoiceNo StockCode                       Description  Quantity  \
89570    543911     21485  RETROSPOT HEART HOT WATER BOTTLE         6   
89571    543911     48138                DOORMAT UNION FLAG         2   
89572    543911     85150     LADIES & GENTLEMEN METAL SIGN        24   
89573    543911     82552               WASHROOM METAL SIGN        24   
89574    543911     82551            LAUNDRY 15C METAL SIGN        12   
...         ...       ...                               ...       ...   
69675    541975     20726                LUNCH BAG WOODLAND        20   
69676    541975     22452      MEASURING TAPE BABUSHKA PINK        24   
69677    541975     22454       MEASURING TAPE BABUSHKA RED        24   
69678    541975     20749        ASSORTED COLOUR MINI CASES         6   
69679    541975     20682  RED RETROSPOT CHILDRENS UMBRELLA         6   

              InvoiceDate  UnitPrice  CustomerID               Country  \
89570 2011-02-14 12:46:00       4.95     17829.0 

<ipython-input-46-1ddecad5b434>:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  China1 = Asia_data[(Asia_data['Date'].dt.month == 1) & (Asia_data['Date'].dt.day >= 15) & (data['Date'].dt.day <= 31)  |


In [47]:
transactions3 = China1.groupby('InvoiceNo')['Description'].apply(list).reset_index()
print(transactions3)

  InvoiceNo                                        Description
0    541975  [4 TRADITIONAL SPINNING TOPS, CHARLIE & LOLA W...
1    542102  [GIN & TONIC DIET GREETING CARD , GREEN REGENC...
2    542276  [LAUNDRY 15C METAL SIGN, BEWARE OF THE CAT MET...
3    543179  [MONEY BOX FIRST ADE DESIGN, MONEY BOX KINGS C...
4    543518  [SET OF 3 NOTEBOOKS IN PARCEL, LUNCH BAG DOLLY...
5    543911  [RETROSPOT HEART HOT WATER BOTTLE, DOORMAT UNI...


In [48]:
transactions3_apply = transactions3['Description'].apply(lambda x: [str(item) for item in x])
transactions3_apply

0    [4 TRADITIONAL SPINNING TOPS, CHARLIE & LOLA W...
1    [GIN & TONIC DIET GREETING CARD , GREEN REGENC...
2    [LAUNDRY 15C METAL SIGN, BEWARE OF THE CAT MET...
3    [MONEY BOX FIRST ADE DESIGN, MONEY BOX KINGS C...
4    [SET OF 3 NOTEBOOKS IN PARCEL, LUNCH BAG DOLLY...
5    [RETROSPOT HEART HOT WATER BOTTLE, DOORMAT UNI...
Name: Description, dtype: object

In [49]:
frequentItemset3 = list(apriori(transactions3_apply, min_support=0.33))

for record in frequentItemset3:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: BASKET OF TOADSTOOLS, LUNCH BAG RED RETROSPOT
Support: 0.3333 (33.33%)
Rule 1: (None -> BASKET OF TOADSTOOLS, LUNCH BAG RED RETROSPOT : Confidence=0.3333, Lift=1.0000)
Rule 2: (BASKET OF TOADSTOOLS -> LUNCH BAG RED RETROSPOT : Confidence=1.0000, Lift=3.0000)
Rule 3: (LUNCH BAG RED RETROSPOT -> BASKET OF TOADSTOOLS : Confidence=1.0000, Lift=3.0000)

yes
Items: BASKET OF TOADSTOOLS, SET 3 RETROSPOT TEA,COFFEE,SUGAR
Support: 0.3333 (33.33%)
Rule 1: (None -> BASKET OF TOADSTOOLS, SET 3 RETROSPOT TEA,COFFEE,SUGAR : Confidence=0.3333, Lift=1.0000)
Rule 2: (BASKET OF TOADSTOOLS -> SET 3 RETROSPOT TEA,COFFEE,SUGAR : Confidence=1.0000, Lift=3.0000)
Rule 3: (SET 3 RETROSPOT TEA,COFFEE,SUGAR -> BASKET OF TOADSTOOLS : Confidence=1.0000, Lift=3.0000)

yes
Items: JAM MAKING SET WITH JARS, BATH BUILDING BLOCK WORD
Support: 0.3333 (33.33%)
Rule 1: (None -> JAM MAKING SET WITH JARS, BATH BUILDING BLOCK WORD : Confidence=0.3333, Lift=1.0000)
Rule 2: (BATH BUILDING BLOCK WORD -> JAM MAKING SET

###ข้อ 4 ช่วงเวลาChristmas & New year ในทวีปยุโรป

In [50]:
#Select 1-31 Dec
December2 = EU_data[(EU_data['Date'].dt.month == 12) & (EU_data['Date'].dt.day >= 1) & (EU_data['Date'].dt.day <= 31)]
#ปริ้นเพื่อตรวจสอบ
print(December2)


       InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

               InvoiceDate  UnitPrice  CustomerID         Country       Dat

In [51]:
#EU Christmas
Eu_christ = EU_data[(EU_data['Date'].dt.month == 12) & (EU_data['Date'].dt.day >= 1) & (EU_data['Date'].dt.day <= 31)]
Eu_christ

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

               InvoiceDate  UnitPrice  CustomerID         Country       Date  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom 2010-12-01  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom 2010-12-01  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
...                    ...        ...         ...             ...        ...  
536964 2011-12-08 10:26:00       5.95     12358.0         Austria 2011-12-08  
536965 2011-12-08 10:26:00       5.95     12358.0         Austria 2011-12-08  
536966 2011-12-08 10:26:00       0.39     12358.0         Austria 2011-12-08  
536967 2011-12-08 10:26:00       0.39     12358.0         Austria 2011-12-08  
536968 2011-12-08 10:26:00      40.00     12358.0         Austria 2011-12-08  

[67852 rows x 9 columns]

In [52]:
transactions4 = Eu_christ.groupby('InvoiceNo')['Description'].apply(list).reset_index()
print(transactions4)

     InvoiceNo                                        Description
0       536365  [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1       536366  [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2       536367  [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3       536368  [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4       536369                         [BATH BUILDING BLOCK WORD]
...        ...                                                ...
3022   C581484                      [PAPER CRAFT , LITTLE BIRDIE]
3023   C581490  [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
3024   C581499                                           [Manual]
3025   C581568                       [VICTORIAN SEWING BOX LARGE]
3026   C581569  [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...

[3027 rows x 2 columns]


In [53]:
transactions4_apply = transactions4['Description'].apply(lambda x: [str(item) for item in x])
transactions4_apply

0       [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1       [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2       [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3       [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4                              [BATH BUILDING BLOCK WORD]
                              ...                        
3022                        [PAPER CRAFT , LITTLE BIRDIE]
3023    [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
3024                                             [Manual]
3025                         [VICTORIAN SEWING BOX LARGE]
3026    [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...
Name: Description, Length: 3027, dtype: object

In [54]:
#items ที่มักซื้อด้วยกัน
frequentItemset4 = list(apriori(transactions4_apply, min_support=0.015))
for record in frequentItemset4:
    if len(record.items) > 2:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")
        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift
            print(f"Rule {i}: ({items_base} -> {items_add}: Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()

yes
Items: PAPER CHAIN KIT 50'S CHRISTMAS , 60 CAKE CASES VINTAGE CHRISTMAS, PAPER CHAIN KIT VINTAGE CHRISTMAS
Support: 0.0182 (1.82%)
Rule 1: (None -> PAPER CHAIN KIT 50'S CHRISTMAS , 60 CAKE CASES VINTAGE CHRISTMAS, PAPER CHAIN KIT VINTAGE CHRISTMAS: Confidence=0.0182, Lift=1.0000)
Rule 2: (60 CAKE CASES VINTAGE CHRISTMAS -> PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS: Confidence=0.4264, Lift=8.9006)
Rule 3: (PAPER CHAIN KIT 50'S CHRISTMAS  -> 60 CAKE CASES VINTAGE CHRISTMAS, PAPER CHAIN KIT VINTAGE CHRISTMAS: Confidence=0.1957, Lift=8.5866)
Rule 4: (PAPER CHAIN KIT VINTAGE CHRISTMAS -> PAPER CHAIN KIT 50'S CHRISTMAS , 60 CAKE CASES VINTAGE CHRISTMAS: Confidence=0.2523, Lift=10.6068)
Rule 5: (PAPER CHAIN KIT 50'S CHRISTMAS , 60 CAKE CASES VINTAGE CHRISTMAS -> PAPER CHAIN KIT VINTAGE CHRISTMAS: Confidence=0.7639, Lift=10.6068)
Rule 6: (60 CAKE CASES VINTAGE CHRISTMAS, PAPER CHAIN KIT VINTAGE CHRISTMAS -> PAPER CHAIN KIT 50'S CHRISTMAS : Confidence=0.7971, Lift=8

In [55]:
frequentItemset4 = list(apriori(transactions4_apply, min_support=0.035))

for record in frequentItemset4:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()

yes
Items: ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKELIKE RED 
Support: 0.0367 (3.67%)
Rule 1: (None -> ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKELIKE RED  : Confidence=0.0367, Lift=1.0000)
Rule 2: (ALARM CLOCK BAKELIKE GREEN -> ALARM CLOCK BAKELIKE RED  : Confidence=0.7070, Lift=14.4602)
Rule 3: (ALARM CLOCK BAKELIKE RED  -> ALARM CLOCK BAKELIKE GREEN : Confidence=0.7500, Lift=14.4602)

yes
Items: CHOCOLATE HOT WATER BOTTLE, HOT WATER BOTTLE I AM SO POORLY
Support: 0.0367 (3.67%)
Rule 1: (None -> CHOCOLATE HOT WATER BOTTLE, HOT WATER BOTTLE I AM SO POORLY : Confidence=0.0367, Lift=1.0000)
Rule 2: (CHOCOLATE HOT WATER BOTTLE -> HOT WATER BOTTLE I AM SO POORLY : Confidence=0.5163, Lift=9.7067)
Rule 3: (HOT WATER BOTTLE I AM SO POORLY -> CHOCOLATE HOT WATER BOTTLE : Confidence=0.6894, Lift=9.7067)

yes
Items: PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS
Support: 0.0479 (4.79%)
Rule 1: (None -> PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS : 

###ข้อ 5 ของวันวาเลนไทน์

In [56]:
#Select 1-31 Dec
Feb14 = data[(data['Date'].dt.month == 2) & (data['Date'].dt.day == 14)]
#ปริ้นเพื่อตรวจสอบ
print(Feb14)

      InvoiceNo StockCode                         Description  Quantity  \
88847    543821     22926      IVORY GIANT GARDEN THERMOMETER         2   
88848    543821     22928     YELLOW GIANT GARDEN THERMOMETER         2   
88849    543821     22925       BLUE GIANT GARDEN THERMOMETER         2   
88850    543821     85175               CACTI T-LIGHT CANDLES        16   
88851    543821     22456     NATURAL SLATE CHALKBOARD LARGE          3   
...         ...       ...                                 ...       ...   
89950    543979     21889              WOODEN BOX OF DOMINOES         1   
89951    543979     85176          SEWING SUSAN 21 NEEDLE SET         1   
89952    543979     37446  MINI CAKE STAND WITH HANGING CAKES         1   
89953    543979     21657     MILK BOTTLE WITH GLASS STOPPER          1   
89954   C543980     22728           ALARM CLOCK BAKELIKE PINK        -1   

              InvoiceDate  UnitPrice  CustomerID         Country       Date  
88847 2011-02-14 09:2

In [57]:
transactions5 = Feb14.groupby('InvoiceNo')['Description'].apply(list).reset_index()
print(transactions5)

   InvoiceNo                                        Description
0     543821  [IVORY GIANT GARDEN THERMOMETER, YELLOW GIANT ...
1     543822  [RETROSPOT TEA SET CERAMIC 11 PC , 12 PENCIL S...
2     543823  [BAKING SET 9 PIECE RETROSPOT , GLASS  BEURRE ...
3     543824  [POPPY'S PLAYHOUSE BEDROOM , POPPY'S PLAYHOUSE...
4     543825  [IVORY GIANT GARDEN THERMOMETER, SET 3 WICKER ...
5     543826  [PURPLE FRANGIPANI HAIRCLIP, FINE SILVER NECKL...
6     543827                    [incorrectly made-thrown away.]
7     543828  [SET OF 6 HERB TINS SKETCHBOOK, SET OF 6 SPICE...
8     543829  [RED RETROSPOT CUP, PINK  POLKADOT CUP, BLUE P...
9     543831  [TOMATO CHARLIE+LOLA COASTER SET, JUMBO BAG TO...
10    543832  [ASSORTED COLOUR BIRD ORNAMENT, 6 RIBBONS RUST...
11    543833  [BROCANTE COAT RACK, SET 3 PAPER VINTAGE CHICK...
12    543836  [SET OF 36 TEATIME PAPER DOILIES, 60 TEATIME F...
13    543839  [FRYING PAN UNION FLAG, RED TOADSTOOL LED NIGH...
14    543842               [GROW A FLYTR

In [58]:
transactions5_apply = transactions5['Description'].apply(lambda x: [str(item) for item in x])
transactions5_apply

0     [IVORY GIANT GARDEN THERMOMETER, YELLOW GIANT ...
1     [RETROSPOT TEA SET CERAMIC 11 PC , 12 PENCIL S...
2     [BAKING SET 9 PIECE RETROSPOT , GLASS  BEURRE ...
3     [POPPY'S PLAYHOUSE BEDROOM , POPPY'S PLAYHOUSE...
4     [IVORY GIANT GARDEN THERMOMETER, SET 3 WICKER ...
5     [PURPLE FRANGIPANI HAIRCLIP, FINE SILVER NECKL...
6                       [incorrectly made-thrown away.]
7     [SET OF 6 HERB TINS SKETCHBOOK, SET OF 6 SPICE...
8     [RED RETROSPOT CUP, PINK  POLKADOT CUP, BLUE P...
9     [TOMATO CHARLIE+LOLA COASTER SET, JUMBO BAG TO...
10    [ASSORTED COLOUR BIRD ORNAMENT, 6 RIBBONS RUST...
11    [BROCANTE COAT RACK, SET 3 PAPER VINTAGE CHICK...
12    [SET OF 36 TEATIME PAPER DOILIES, 60 TEATIME F...
13    [FRYING PAN UNION FLAG, RED TOADSTOOL LED NIGH...
14                 [GROW A FLYTRAP OR SUNFLOWER IN TIN]
15    [CHARLOTTE BAG SUKI DESIGN, JUMBO STORAGE BAG ...
16    [CARRIAGE, CLASSIC METAL BIRDCAGE PLANT HOLDER...
17    [CHARLOTTE BAG SUKI DESIGN, WOODLAND CHARL

In [59]:

frequentItemset5 = list(apriori(transactions5_apply, min_support=0.09))

for record in frequentItemset5:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: SET OF 3 CAKE TINS PANTRY DESIGN , HOMEMADE JAM SCENTED CANDLES
Support: 0.0909 (9.09%)
Rule 1: (None -> SET OF 3 CAKE TINS PANTRY DESIGN , HOMEMADE JAM SCENTED CANDLES : Confidence=0.0909, Lift=1.0000)
Rule 2: (HOMEMADE JAM SCENTED CANDLES -> SET OF 3 CAKE TINS PANTRY DESIGN  : Confidence=1.0000, Lift=6.8750)
Rule 3: (SET OF 3 CAKE TINS PANTRY DESIGN  -> HOMEMADE JAM SCENTED CANDLES : Confidence=0.6250, Lift=6.8750)

yes
Items: SET OF 3 CAKE TINS PANTRY DESIGN , SET OF 6 SPICE TINS PANTRY DESIGN
Support: 0.0909 (9.09%)
Rule 1: (None -> SET OF 3 CAKE TINS PANTRY DESIGN , SET OF 6 SPICE TINS PANTRY DESIGN : Confidence=0.0909, Lift=1.0000)
Rule 2: (SET OF 3 CAKE TINS PANTRY DESIGN  -> SET OF 6 SPICE TINS PANTRY DESIGN : Confidence=0.6250, Lift=5.7292)
Rule 3: (SET OF 6 SPICE TINS PANTRY DESIGN -> SET OF 3 CAKE TINS PANTRY DESIGN  : Confidence=0.8333, Lift=5.7292)



###ข้อ6 จัดกลุ่มตามประเทศและเดือน

In [110]:
data['InvoiceDate']=pd.to_datetime(data.InvoiceDate, format='%m/%d/%Y %H:%M' )
data

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country       Date  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom 2010-12-01  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom 2010-12-01  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
...                    ...        ...         ...             ...        ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France 2011-12-09  
541905 2011-12-09 12:50:00       2.10     12680.0          France 2011-12-09  
541906 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541907 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541908 2011-12-09 12:50:00       4.95     12680.0          France 2011-12-09  

[541909 rows x 9 columns]

In [112]:
data['Month'] = data['InvoiceDate'].dt.month
data

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country       Date  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom 2010-12-01   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom 2010-12-01   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01   
...                    ...        ...         ...             ...        ...   
541904 2011-12-09 12:50:00       0.85     12680.0          France 2011-12-09   
541905 2011-12-09 12:50:00       2.10     12680.0          France 2011-12-09   
541906 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09   
541907 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09   
541908 2011-12-09 12:50:00       4.95     12680.0          France 2011-12-09   

        Month  
0          12  
1          12  
2          12  
3          12  
4          12  
...       ...  
541904     12  
541905     12  
541906     12  
541907     12  
541908     12  

[541909 rows x 10 columns]

In [119]:
transWorld = data.groupby(['InvoiceNo','Country', 'Month'])['StockCode'].apply(list)
transWorld

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


InvoiceNo  Country         Month
536365     United Kingdom  12       [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366     United Kingdom  12                                          [22633, 22632]
536367     United Kingdom  12       [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368     United Kingdom  12                            [22960, 22913, 22912, 22914]
536369     United Kingdom  12                                                 [21756]
                                                          ...                        
C581484    United Kingdom  12                                                 [23843]
C581490    United Kingdom  12                                          [22178, 23144]
C581499    United Kingdom  12                                                     [M]
C581568    United Kingdom  12                                                 [21258]
C581569    United Kingdom  12                                          [84978, 20979]
Name: StockCode, Leng

In [123]:
min_sup = 0.025
frequentItemsets6 = list(apriori(transWorld,min_support = min_sup))
for i in frequentItemsets6:
    if i[0].__len__()>1:
        print('yes')
        print(i)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


yes
RelationRecord(items=frozenset({'20727', '20725'}), support=0.025019305019305018, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20727', '20725'}), confidence=0.025019305019305018, lift=1.0), OrderedStatistic(items_base=frozenset({'20725'}), items_add=frozenset({'20727'}), confidence=0.40298507462686567, lift=8.059701492537313), OrderedStatistic(items_base=frozenset({'20727'}), items_add=frozenset({'20725'}), confidence=0.5003861003861003, lift=8.059701492537313)])
yes
RelationRecord(items=frozenset({'20725', '22383'}), support=0.0255984555984556, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725', '22383'}), confidence=0.0255984555984556, lift=1.0), OrderedStatistic(items_base=frozenset({'20725'}), items_add=frozenset({'22383'}), confidence=0.41231343283582095, lift=8.176813101414826), OrderedStatistic(items_base=frozenset({'22383'}), items_add=frozenset({'20725'}), confidence=0.5076569678407351, lift=8.176

In [125]:
set(data[(data['StockCode']=='85099B') | (data['StockCode']=='22386')]['Description'])


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


{'JUMBO BAG PINK POLKADOT', 'JUMBO BAG RED RETROSPOT'}